src/develop/imageop.c

```diff
@@ -1353,38 +1348,76 @@ void dt_iop_gui_update_header(dt_iop_module_t *module)
+    GList *children = gtk_container_get_children(GTK_CONTAINER(gtk_widget_get_parent(module->widget)));
+    label_widget = g_list_nth_data(children, 0);
+    g_list_free(children);
+    if(strcmp(gtk_widget_get_name(label_widget), "iop-plugin-warning"))
+      label_widget = NULL;
   }
 
-  if(!module) return;
+  if(trouble_msg && *trouble_msg)
+  {
+    if((!module || !module->has_trouble) && (stderr_message || !module->widget))
+    {
+      const char *name = module ? module->name() : "?";
+      fprintf(stderr,"[%s] %s\n", name, stderr_message ? stderr_message : trouble_msg);
+    }
 
-  dt_iop_gui_enter_critical_section(module);
+    if(module && module->widget)
+    {
+      if(label_widget)
+      {
+        // set the warning message in the module's message area just below the header
+        gtk_label_set_text(GTK_LABEL(label_widget), trouble_msg);
+      }
+      else
+      {
+        label_widget = gtk_label_new(trouble_msg);;
+        gtk_label_set_line_wrap(GTK_LABEL(label_widget), TRUE);
+        gtk_label_set_xalign(GTK_LABEL(label_widget), 0.0);
+        gtk_widget_set_name(label_widget, "iop-plugin-warning");
+
+        GtkWidget *iopw = gtk_widget_get_parent(module->widget);
+        gtk_box_pack_start(GTK_BOX(iopw), label_widget, TRUE, TRUE, 0);
+        gtk_box_reorder_child(GTK_BOX(iopw), label_widget, 0);
+        gtk_widget_show(label_widget);
+      }
 
-  // convert empty string to NULL for easier handling
-  if(trouble_msg && !(*trouble_msg)) trouble_msg = NULL;
-  if(trouble_tooltip && !(*trouble_tooltip)) trouble_tooltip = NULL;
+      gtk_widget_set_tooltip_text(GTK_WIDGET(label_widget), trouble_tooltip);
 
-  // check if anything has changed. only consider tooltip if message is not empty
-  if(g_strcmp0(trouble_msg, module->trouble_message) == 0
-     && (!trouble_msg || g_strcmp0(trouble_tooltip, module->trouble_tooltip) == 0))
-  {
-    dt_iop_gui_leave_critical_section(module);
-    return;
+      // set the module's trouble flag
+      module->has_trouble = TRUE;
+
+      _iop_gui_update_header(module);
+    }
   }
+  else if(module && module->has_trouble)
+  {
+    // no more trouble, so clear the trouble flag and remove the message area
+    module->has_trouble = FALSE;
+
+    _iop_gui_update_header(module);
 
-  g_free(module->trouble_message);
-  g_free(module->trouble_tooltip);
-  module->trouble_message = g_strdup(trouble_msg);
-  module->trouble_tooltip = g_strdup(trouble_tooltip);
-  module->trouble_message_updated = TRUE;
+    if(label_widget) gtk_widget_destroy(label_widget);
+  }
+}
 
-  dt_iop_gui_leave_critical_section(module);
+void dt_iop_set_module_trouble_message(dt_iop_module_t *const module,
+                                       char* const trouble_msg, const char* const trouble_tooltip,
+                                       const char *const stderr_message)
+{
+  DT_DEBUG_CONTROL_SIGNAL_RAISE(darktable.signals, DT_SIGNAL_TROUBLE_MESSAGE,
```

Comment:
The only worry I have here is whether the strings are actually stored in the signal that is being sent or just a pointer to them. Might not be a problem now, as only constant strings are used, but if someone temporarily allocates a string, passes it here and then frees on return, I assume it'll crash? Sorry, can't test now.

Similarly, can we assume that module still exists when we reach the callback function? It must work, otherwise I guess color picker also would have problems, but I can't say I understand _why_ it works ;-)

Reply:
Having an additional thought; from the description of g_main_context_invoke, used in dt_control_signal_raise, I understand that signals are sent as idle event on the main loop, with priority G_PRIORITY_DEFAULT_IDLE. Maybe for this case, GTK_PRIORITY_RESIZE would be more appropriate, since adding/removing the label would trigger a resize/redraw and could cause a slight flicker or delay. Probably not worth having a whole separate version of DT_DEBUG_CONTROL_SIGNAL_RAISE etc for. But maybe experimenting with a higher priority for _all_ signals used in dt could be interesting...

Reply:
> The only worry I have here is whether the strings are actually stored in the signal that is being sent or just a pointer to them. Might not be a problem now, as only constant strings are used, but if someone temporarily allocates a string, passes it here and then frees on return, I assume it'll crash? Sorry, can't test now.

Yes, it will crash. We have two options:
- add a message that currently the strings must be static
- always take a copy of the string and free them when displayed

Reply:
> * always take a copy of the string and free them when displayed

This https://github.com/darktable-org/darktable/blob/ad87231986fa519d2d30e49a8f6c11ab3e44d91c/src/control/signal.c#L321 suggests there is a way to do this automatically (have the signal own GValues that contain the strings). Just changing the type in your marshalling definition from G_TYPE_POINTER to G_TYPE_STRING might do it?

Reply:
Maybe, no expert here and the doc doesn't help :(

Reply:
When you say "doc", do you mean "source"? ;-)

Reply:
No, the Gtk+ doc, not clear to me.

Reply:
https://developer.gnome.org/gobject/stable/gobject-Standard-Parameter-and-Value-Types.html#g-value-set-string says "caller-owned string to be duplicated for the GValue.", so I assume it will be freed correctly here: https://github.com/darktable-org/darktable/blob/ad87231986fa519d2d30e49a8f6c11ab3e44d91c/src/control/signal.c#L236

Other 2 alternatives are always using static string (error-prone) or creating a destructor for a signal freeing the string (which is just a contrived way to do the same). So using G_TYPE_STRING is IMHO correct way to do it.

Reply:
@parafin : Thanks, I have changed to G_TYPE_STRING so I think we are all set now.